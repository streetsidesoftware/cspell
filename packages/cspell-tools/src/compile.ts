import { writeFile } from 'node:fs/promises';

import { opConcatMap, pipe } from '@cspell/cspell-pipe/sync';
import YAML from 'yaml';

import type { CompileCommonAppOptions } from './AppOptions.js';
import { compile } from './compiler/compile.js';
import { createCompileRequest } from './compiler/createCompileRequest.js';
import type { CompileRequest } from './config/config.js';
import type { FeatureFlags } from './FeatureFlags/index.js';
import { getSystemFeatureFlags, parseFlags } from './FeatureFlags/index.js';
import { globP } from './util/globP.js';

getSystemFeatureFlags().register('compound', 'Enable compound dictionary sources.');

const defaultConfigFile = 'cspell-tools.config.yaml';

export async function processCompileAction(
    src: string[],
    options: CompileCommonAppOptions,
    featureFlags: FeatureFlags | undefined
): Promise<void> {
    const ff = featureFlags || getSystemFeatureFlags();
    parseFlags(ff, options.experimental || []);
    return useCompile(src, options);
}

async function useCompile(src: string[], options: CompileCommonAppOptions): Promise<void> {
    console.log(
        'Compile:\n output: %s\n compress: %s\n files:\n  %s ',
        options.output || 'default',
        options.compress ? 'true' : 'false',
        src.join('\n  ')
    );
    if (options.listFile && options.listFile.length) {
        console.log(' list files:\n  %s', options.listFile.join('\n  '));
    }
    console.log('\n\n');

    const globResults = await Promise.all(src.map((s) => globP(s)));
    const sources = [
        ...pipe(
            globResults,
            opConcatMap((a) => a)
        ),
    ];

    const request = createCompileRequest(sources, options);
    return options.init ? initConfig(request) : compile(request);
}

async function initConfig(request: CompileRequest): Promise<void> {
    const content = YAML.stringify(request, null, 2);
    console.log('Writing config file: %s', defaultConfigFile);
    await writeFile(defaultConfigFile, content);

    console.log(`Init complete.
To build, use:
  cspell-tools-cli build
`);
}
