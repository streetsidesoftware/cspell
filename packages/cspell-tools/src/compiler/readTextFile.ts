import * as fs from 'fs-extra';
import * as zlib from 'zlib';

export function readTextFile(filename: string): Promise<string> {
    const content = fs
        .readFile(filename)
        .then((buffer) => (/\.gz$/.test(filename) ? zlib.gunzipSync(buffer) : buffer))
        .then((buffer) => buffer.toString('utf8'));
    return content;
}

export async function readTextFileLines(filename: string): Promise<string[]> {
    const content = await readTextFile(filename);
    return content.split('\n');
}
