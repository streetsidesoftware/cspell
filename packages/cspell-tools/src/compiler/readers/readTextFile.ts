import { promises as fs } from 'fs';
import { promisify } from 'util';
import * as zlib from 'zlib';

const gunzip = promisify(zlib.gunzip);

const isGzFile = /\.gz$/;

export function readTextFile(filename: string): Promise<string> {
    const content = fs
        .readFile(filename)
        .then((buffer) => (isGzFile.test(filename) ? gunzip(buffer) : buffer))
        .then((buffer) => buffer.toString('utf8'));
    return content;
}

export async function readTextFileLines(filename: string): Promise<string[]> {
    const content = await readTextFile(filename);
    return content.split('\n');
}
