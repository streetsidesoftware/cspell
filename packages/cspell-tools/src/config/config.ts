export interface RunConfig extends Partial<CompileRequest> {
    /**
     * Url to JSON Schema
     * @default "https://raw.githubusercontent.com/streetsidesoftware/cspell/main/packages/cspell-tools/cspell-tools.config.schema.json"
     */
    $schema?: string;

    /**
     * Optional Target Dictionaries to create.
     */
    targets?: Target[];
    /**
     * Specify the directory where all relative paths will resolved against.
     * By default, all relative paths are relative to the location of the
     * config file.
     */
    rootDir?: string;
}

export interface CompileRequest extends CompileTargetOptions, CompileSourceOptions {
    /**
     * Specify the directory where all relative paths will resolved against.
     * By default, all relative paths are relative to the current directory.
     */
    rootDir?: string;

    /**
     * Target Dictionaries to create.
     */
    targets: Target[];

    /**
     * Path to checksum file. `true` - defaults to `./checksum.txt`.
     */
    checksumFile?: string | boolean;
}

export interface Experimental {
    /**
     * Experimental flags
     */
    experimental?: string[] | undefined;
}

export interface CompileTargetOptions {
    /**
     * Generate lower case / accent free versions of words.
     * @default true
     */
    generateNonStrict?: boolean | undefined;

    /**
     * Sort the words in the resulting dictionary.
     * Does not apply to `trie` based formats.
     * @default: true
     */
    sort?: boolean | undefined;

    /**
     * Words in the `allowedSplitWords` are considered correct and can be used
     * as a basis for splitting compound words.
     *
     * If entries can be split so that all the words in the entry are allowed,
     * then only the individual words are added, otherwise the entire entry is added.
     * This is to prevent misspellings in CamelCase words from being introduced into the
     * dictionary.
     */
    allowedSplitWords?: FilePath | FilePath[] | undefined;
}

export interface Target extends CompileTargetOptions {
    /**
     * Name of target, used as the basis of target file name.
     */
    name: string;

    /**
     * The target directory
     * @default current directory
     */
    targetDirectory?: FilePath | undefined;

    /**
     * gzip the file?
     * @default: false
     */
    compress?: boolean | undefined;

    /**
     * Format of the dictionary.
     */
    format: DictionaryFormats;

    /**
     * File sources used to build the dictionary.
     */
    sources: DictionarySource[];

    /**
     * Words from the sources that are found in `excludeWordsFrom` files
     * will not be added to the dictionary.
     *
     * @version 8.3.2
     */
    excludeWordsFrom?: FilePath[] | undefined;

    /**
     * Advanced: Set the trie base number. A value between 10 and 36
     * Set numeric base to use.
     * 10 is the easiest to read.
     * 16 is common hex format.
     * 36 is the most compact.
     */
    trieBase?: number | undefined;
}

export type DictionaryFormats = 'plaintext' | 'trie' | 'trie3' | 'trie4';

/**
 * Note: All relative paths are relative to the config file location.
 */
export type FilePath = string;

export type DictionarySource = FilePath | FileSource | FileListSource;

export interface FileSource extends CompileSourceOptions {
    filename: FilePath;
}

export interface FileListSource extends CompileSourceOptions {
    listFile: FilePath;
}

export interface CompileSourceOptions {
    /**
     * Maximum number of nested Hunspell Rules to apply.
     * This is needed for recursive dictionaries like Hebrew.
     */
    maxDepth?: number | undefined;
    /**
     * Split lines into words.
     * @default false
     */
    split?: boolean | 'legacy' | undefined;
    /**
     * Do not generate lower case / accent free versions of words.
     * @default false
     */
    keepRawCase?: boolean | undefined;

    allowedSplitWords?: FilePath | FilePath[] | undefined;
}

export const configFileSchemaURL =
    'https://raw.githubusercontent.com/streetsidesoftware/cspell/main/packages/cspell-tools/cspell-tools.config.schema.json';
