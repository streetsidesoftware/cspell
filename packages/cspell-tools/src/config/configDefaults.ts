import type { RequireFields } from '../types.js';
import type { CompileSourceOptions } from './config.js';

export const defaultCompileSourceOptions = {
    maxDepth: undefined,
    split: false,
    keepRawCase: false,
    allowedSplitWords: undefined,
    storeSplitWordsAsCompounds: false,
    minCompoundLength: 4,
} as const satisfies RequireFields<CompileSourceOptions>;
