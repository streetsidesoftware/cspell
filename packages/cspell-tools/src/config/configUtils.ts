import type { DictionarySource, FileListSource, FilePath, FileSource } from '../config';

export function isFilePath(source: DictionarySource): source is FilePath {
    return typeof source === 'string';
}

export function isFileSource(source: DictionarySource): source is FileSource {
    if (!source || isFilePath(source)) return false;
    return (<FileSource>source).filename !== undefined;
}

export function isFileListSource(source: DictionarySource): source is FileListSource {
    if (!source || isFilePath(source)) return false;
    return (<FileListSource>source).listFile !== undefined;
}
