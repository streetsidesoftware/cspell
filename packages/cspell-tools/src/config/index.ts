export {
    CompileRequest,
    CompileTargetOptions,
    DictionaryFormats,
    DictionarySource,
    FileListSource,
    FilePath,
    FileSource,
    RunConfig,
    SourceConfig,
    Target,
} from './config';
export { normalizeConfig } from './normalizeConfig';
