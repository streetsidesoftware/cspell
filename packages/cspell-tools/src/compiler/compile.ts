import { pipeAsync, toArray } from '@cspell/cspell-pipe';
import { opAwaitAsync, opMapAsync } from '@cspell/cspell-pipe/operators';
import { opConcatMap, opMap, pipe } from '@cspell/cspell-pipe/sync';
import * as path from 'path';
import type {
    CompileRequest,
    CompileSourceOptions,
    CompileTargetOptions,
    DictionarySource,
    FilePath,
    FileSource,
    Target,
} from '../config';
import { isFileListSource, isFilePath, isFileSource } from '../config';
import { streamWordsFromFile } from './iterateWordsFromFile';
import { logWithTimestamp } from './logWithTimestamp';
import type { ReaderOptions } from './Reader';
import { readTextFile } from './readTextFile';
import { compileTrie, compileWordList } from './wordListCompiler';
import { normalizeTargetWords } from './wordListParser';

interface CompileOptions {
    /**
     * Optional filter function to filter targets.
     */
    filter?: (target: Target) => boolean;
}

export async function compile(request: CompileRequest, options?: CompileOptions): Promise<void> {
    const { targets } = request;

    // console.log('Request: %o', request);

    const rootDir = path.resolve(request.rootDir || '.');
    const targetOptions: CompileTargetOptions = {
        sort: request.sort,
        generateNonStrict: request.generateNonStrict,
    };

    for (const target of targets) {
        const keep = options?.filter?.(target) ?? true;
        if (!keep) continue;
        const adjustedTarget: Target = { ...targetOptions, ...target };
        await compileTarget(adjustedTarget, request, rootDir);
    }
    logWithTimestamp(`Complete.`);
}

export async function compileTarget(target: Target, options: CompileSourceOptions, rootDir: string): Promise<void> {
    logWithTimestamp(`Start compile: ${target.name}`);

    // console.log('Target: %o', target);

    const { format, sources, trieBase, sort = true, generateNonStrict = false } = target;
    const targetDirectory = path.resolve(rootDir, target.targetDirectory ?? process.cwd());
    const generateNonStrictTrie = target.generateNonStrict ?? true;

    const name = normalizeTargetName(target.name);

    const useTrie = format.startsWith('trie');
    const filename = resolveTarget(name, targetDirectory, useTrie, target.compress ?? false);

    const filesToProcessAsync = pipeAsync(
        readSourceList(sources, rootDir),
        opMapAsync((src) => readFileSource(src, options)),
        opAwaitAsync()
    );
    const filesToProcess: FileToProcess[] = await toArray(filesToProcessAsync);
    const normalizer = normalizeTargetWords({ sort: useTrie || sort, generateNonStrict });

    const action = useTrie
        ? async (words: Iterable<string>, dst: string) => {
              return compileTrie(pipe(words, normalizer), dst, {
                  base: trieBase,
                  sort: false,
                  trie3: format === 'trie3',
                  trie4: format === 'trie4',
                  generateNonStrict: generateNonStrictTrie,
              });
          }
        : async (words: Iterable<string>, dst: string) => {
              return compileWordList(pipe(words, normalizer), dst, { sort, generateNonStrict });
          };

    await processFiles(action, filesToProcess, filename);

    logWithTimestamp(`Done compile: ${target.name}`);
}

function rel(filePath: string): string {
    return path.relative(process.cwd(), filePath);
}

async function processFiles(action: ActionFn, filesToProcess: FileToProcess[], mergeTarget: string) {
    const toProcess = filesToProcess;
    const dst = mergeTarget;

    const words = pipe(
        toProcess,
        opMap((ftp) => {
            const { src } = ftp;
            logWithTimestamp('Process "%s" to "%s"', rel(src), rel(dst));
            return ftp;
        }),
        opConcatMap(function* (ftp) {
            yield* ftp.words;
            logWithTimestamp('Done processing %s', rel(ftp.src));
        })
        // opMap((a) => (console.warn(a), a))
    );
    await action(words, dst);
    logWithTimestamp('Done "%s"', rel(dst));
}
interface FileToProcess {
    src: string;
    words: Iterable<string>;
}

type ActionFn = (words: Iterable<string>, dst: string) => Promise<void>;

function resolveTarget(name: string, directory: string, useTrie: boolean, useGzCompress: boolean | boolean): string {
    const ext = ((useTrie && '.trie') || '.txt') + ((useGzCompress && '.gz') || '');
    const filename = name + ext;
    return path.resolve(directory, filename);
}

function readSourceList(sources: DictionarySource[], rootDir: string): AsyncIterable<FileSource> {
    async function* mapSrc(): AsyncIterable<FileSource> {
        for (const src of sources) {
            if (isFilePath(src)) {
                yield { filename: path.resolve(rootDir, src) };
                continue;
            }
            if (isFileSource(src)) {
                yield { ...src, filename: path.resolve(rootDir, src.filename) };
                continue;
            }
            if (isFileListSource(src)) {
                const { listFile, ...rest } = src;
                const absListFile = path.resolve(rootDir, listFile);
                const listFileDir = path.dirname(absListFile);
                const files = await readFileList(absListFile);
                for (const filename of files) {
                    yield { ...rest, filename: path.resolve(listFileDir, filename) };
                }
            }
        }
    }

    return mapSrc();
}

async function readFileList(fileList: FilePath): Promise<string[]> {
    const content = await readTextFile(fileList);
    return content
        .split('\n')
        .map((a) => a.trim())
        .filter((a) => !!a);
}

async function readFileSource(fileSource: FileSource, sourceOptions: CompileSourceOptions): Promise<FileToProcess> {
    const {
        filename,
        keepRawCase = sourceOptions.keepRawCase || false,
        split = sourceOptions.split || false,
        maxDepth,
    } = fileSource;

    const legacy = split === 'legacy';
    const splitWords = legacy ? false : split;

    // console.warn('fileSource: %o,\n targetOptions %o, \n opt: %o', fileSource, targetOptions, opt);

    const readerOptions: ReaderOptions = { maxDepth, legacy, splitWords, keepCase: keepRawCase };

    logWithTimestamp(`Reading ${path.basename(filename)}`);
    const stream = await streamWordsFromFile(filename, readerOptions);
    logWithTimestamp(`Done reading ${path.basename(filename)}`);
    const f: FileToProcess = {
        src: filename,
        words: stream,
    };
    return f;
}

function normalizeTargetName(name: string) {
    return name.replace(/((\.txt|\.dic|\.aff|\.trie)(\.gz)?)?$/, '').replace(/[^\p{L}\p{M}.\w\\/-]/gu, '_');
}
