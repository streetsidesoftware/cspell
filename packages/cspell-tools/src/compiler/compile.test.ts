import * as path from 'path';
import { spyOnConsole } from '../test/console';
import { createTestHelper } from '../test/TestHelper';
import { compile } from './compile';
import { CompileRequest, Target } from './config';
import { readTextFile } from './readTextFile';

const testHelper = createTestHelper(__filename);

const pathSamples = path.join(testHelper.packageRoot, '../Samples/dicts');

function sample(...parts: string[]): string {
    return path.resolve(pathSamples, ...parts);
}

spyOnConsole();

describe('compile', () => {
    beforeAll(() => {
        testHelper.clearTempDir();
    });

    beforeEach(() => {
        testHelper.cdToTempDir();
    });

    test.each`
        format         | compress
        ${'plaintext'} | ${false}
        ${'plaintext'} | ${true}
    `('compile', async ({ format, compress }) => {
        const target: Target = {
            filename: 'myDictionary',
            format,
            sources: [sample('cities.txt')],
            compress,
        };
        const req: CompileRequest = {
            targets: [target],
        };

        await compile(req);

        const content = await readTextFile('myDictionary.txt');
        expect(content).toMatchSnapshot();
    });
});
