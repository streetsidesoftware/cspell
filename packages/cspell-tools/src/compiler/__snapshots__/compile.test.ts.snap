// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile compile 1`] = `
"
# cspell-tools: keep-case no-split

London
Los Angeles
Mexico City
New Amsterdam
New Delhi
New York
Paris
San Francisco
~london
~los angeles
~mexico city
~new amsterdam
~new delhi
~new york
~paris
~san francisco
"
`;

exports[`compile compile 2`] = `
"
# cspell-tools: keep-case no-split

London
Los Angeles
Mexico City
New Amsterdam
New Delhi
New York
Paris
San Francisco
~london
~los angeles
~mexico city
~new amsterdam
~new delhi
~new york
~paris
~san francisco
"
`;
